compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala

```diff
@@ -952,39 +1069,82 @@ class JSCodeGen()(using genCtx: Context) {
    *  an explicit parameter for their `this` value.
    */
   private def genMethodDef(namespace: js.MemberNamespace, methodName: js.MethodIdent,
-      originalName: OriginalName,paramsSyms: List[Symbol], resultIRType: jstpe.Type,
+      originalName: OriginalName, paramsSyms: List[Symbol], resultIRType: jstpe.Type,
       tree: Tree, optimizerHints: OptimizerHints): js.MethodDef = {
     implicit val pos = tree.span
 
-    val jsParams = for (param <- paramsSyms) yield {
-      implicit val pos = param.span
-      js.ParamDef(encodeLocalSym(param), originalNameOfLocal(param),
-          toIRType(param.info), mutable = false, rest = false)
-    }
+    val jsParams = paramsSyms.map(genParamDef(_))
 
     def genBody() = localNames.makeLabeledIfRequiresEnclosingReturn(resultIRType) {
       if (resultIRType == jstpe.NoType) genStat(tree)
       else genExpr(tree)
     }
 
-    //if (!isScalaJSDefinedJSClass(currentClassSym)) {
-    val flags = js.MemberFlags.empty.withNamespace(namespace)
-    js.MethodDef(flags, methodName, originalName, jsParams, resultIRType, Some(genBody()))(
-          optimizerHints, None)
-    /*} else {
+    if (!currentClassSym.isNonNativeJSClass) {
+      val flags = js.MemberFlags.empty.withNamespace(namespace)
+      js.MethodDef(flags, methodName, originalName, jsParams, resultIRType, Some(genBody()))(
+            optimizerHints, None)
+    } else {
       assert(!namespace.isStatic, tree.span)
 
+      val thisLocalIdent = freshLocalIdent("this")
       withScopedVars(
-        thisLocalVarIdent := Some(freshLocalIdent("this"))
+        thisLocalVarIdent := Some(thisLocalIdent)
       ) {
-        val thisParamDef = js.ParamDef(thisLocalVarIdent.get.get,
+        val staticNamespace =
+          if (namespace.isPrivate) js.MemberNamespace.PrivateStatic
+          else js.MemberNamespace.PublicStatic
+        val flags =
+          js.MemberFlags.empty.withNamespace(staticNamespace)
+        val thisParamDef = js.ParamDef(thisLocalIdent, thisOriginalName,
             jstpe.AnyType, mutable = false, rest = false)
 
-        js.MethodDef(static = true, methodName, thisParamDef :: jsParams,
-            resultIRType, genBody())(
+        js.MethodDef(flags, methodName, originalName,
+            thisParamDef :: jsParams, resultIRType, Some(genBody()))(
             optimizerHints, None)
       }
-    }*/
+    }
+  }
+
+  /** Moves all statements after the super constructor call.
+   *
+   *  This is used for the primary constructor of a non-native JS class,
+   *  because those cannot access `this` before the super constructor call.
+   *
+   *  dotc inserts statements before the super constructor call for param
+   *  accessor initializers (including val's and var's declared in the params).
+   *  We move those after the super constructor call, and are therefore
+   *  executed later than for a Scala class.
```

Comment:
Yes. That semantic difference already exists in Scala 2, and is unavoidable.

Reply:
I have expanded the comment to make it clear that it is related to Scala/Scala.js as *languages*, rather than a workaround for any particular *compiler*.